fili-core/src/main/java/com/yahoo/bard/webservice/web/filterparser/ApiFilterListener.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.filterparser;
+
+import java.util.Set;
+
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionDictionary;
+import com.yahoo.bard.webservice.data.dimension.DimensionField;
+import com.yahoo.bard.webservice.table.LogicalTable;
+import com.yahoo.bard.webservice.web.BadApiRequestException;
+import com.yahoo.bard.webservice.web.BadFilterException;
+import com.yahoo.bard.webservice.web.FilterOperation;
+import com.yahoo.bard.webservice.web.filterparser.FiltersParser.FilterContext;
+
+/**
+ * Listener to construct single ApiFilter objects.
+ */
+public class ApiFilterListener extends ApiFiltersListListener {
+    private Dimension dimension = null;
+    private DimensionField dimensionField = null;
+    private FilterOperation operation = null;
+    private Set<String> values = null;
+
+    /**
+     * Construct the listener.
+     *
+     * @param table LogicalTable to use for the filter
+     * @param dimensionDictionary Dictionary of dimensions
+     */
+    public ApiFilterListener(LogicalTable table, DimensionDictionary dimensionDictionary) {
+        super(table, dimensionDictionary);
+    }
+
+    /* (non-Javadoc)
+     * @see com.yahoo.bard.webservice.web.filterparser.ApiFiltersListListener#exitFilter(
+     * com.yahoo.bard.webservice.web.filterparser.FiltersParser.FilterContext)
+     */
+    @Override
+    public void exitFilter(FilterContext ctx) {
+        FiltersParser.FilterComponentContext f = ctx.filterComponent();
+        dimension = extractDimension(f);
+        if (dimension == null) {
+            return;
+        }
+
+        dimensionField = extractField(f, dimension);
+        if (dimensionField == null) {
+            return;
+        }
+
+        operation = extractOperation(f);
+        if (operation == null) {
+            return;
+        }
+
+        values = extractValues(f);
+    }
+
+    /**
+     * Simple getter for the Dimension that also checks for any pending errors and throws exceptions for them.
+     * @return the Dimension
+     * @throws BadFilterException invalid filter expression
+     * @throws BadApiRequestException filter expression that doesn't match the current table or dimensions
+     */
+    public Dimension getDimension() throws BadFilterException, BadApiRequestException {
+        processErrors();
```

Comment:
I'm not a particularly big fan of this "delay on error reporting until we actually try to use the object." If we want that, we should just ditch the whole type system and go duck ;). 

Is there a reason we can't call `processErrors` in the `exitFilter` method where we build out the listener's fields?